# Regressão 01 - tarefa 02 - transformações em X e Y

### 1. Melhorando o ajuste no percentual de gorjetas

a. Vamos partir do modelo feito na última tarefa, o que relaciona ```tip_pct``` e ```net_bill```. Carregue a base, os pacotes e reproduza este modelo aqui.

In [10]:
import pandas as pd
import seaborn as sns
from seaborn import load_dataset

import matplotlib.pyplot as plt

import numpy as np

import statsmodels.formula.api as smf
import statsmodels.api as sm

from sklearn.preprocessing import PolynomialFeatures

In [11]:
tips_df = sns.load_dataset("tips")
tips_df['tip_pct'] = tips_df['tip'] / (tips_df['total_bill'] - tips_df['tip'])
tips_df['net_bill'] = tips_df['total_bill'] - tips_df['tip']
tips_df.head()

total_bill   tip     sex smoker  day    time  size   tip_pct  net_bill
0       16.99  1.01  Female     No  Sun  Dinner     2  0.063204     15.98
1       10.34  1.66    Male     No  Sun  Dinner     3  0.191244      8.68
2       21.01  3.50    Male     No  Sun  Dinner     3  0.199886     17.51
3       23.68  3.31    Male     No  Sun  Dinner     2  0.162494     20.37
4       24.59  3.61  Female     No  Sun  Dinner     4  0.172069     20.98

In [12]:
X_pct = tips_df['net_bill']
y_pct = tips_df['tip_pct']
X_pct = sm.add_constant(X_pct)

model_pct = sm.OLS(y_pct, X_pct).fit()

model_pct_summary = model_pct.summary()
model_pct_summary

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                tip_pct   R-squared:                       0.091
Model:                            OLS   Adj. R-squared:                  0.088
Method:                 Least Squares   F-statistic:                     24.31
Date:                Mon, 06 May 2024   Prob (F-statistic):           1.52e-06
Time:                        21:17:36   Log-Likelihood:                 108.00
No. Observations:                 244   AIC:                            -212.0
Df Residuals:                     242   BIC:                            -205.0
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          0.3053      0.023     13.166      0.000       0.260       0.351
net_bill      -0.0061      0.001     -4.931      0.000      -0.009      -0.004
==============================================================================
Omnibus:                      468.674   Durbin-Watson:                   1.987
Prob(Omnibus):                  0.000   Jarque-Bera (JB):           234437.604
Skew:                          11.104   Prob(JB):                         0.00
Kurtosis:                     153.221   Cond. No.                         43.3
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

### 2. ajuste o modelo de ```tip_pct``` em função do logaritmo de ```net_bill```.

In [13]:
import numpy as np

# Preparando os dados para a regressão com o logaritmo de 'net_bill'
X_log = np.log(tips_df['net_bill'])  # aplicando logaritmo
y = tips_df['tip_pct']               # variável dependente

# Adicionando a constante para o intercepto
X_log = sm.add_constant(X_log)

# Criando o modelo de regressão linear com logaritmo
model_log = sm.OLS(y, X_log)

# Ajustando o modelo
results_log = model_log.fit()

# Exibindo os resultados
results_log.summary()


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                tip_pct   R-squared:                       0.202
Model:                            OLS   Adj. R-squared:                  0.198
Method:                 Least Squares   F-statistic:                     61.16
Date:                Mon, 06 May 2024   Prob (F-statistic):           1.62e-13
Time:                        21:17:36   Log-Likelihood:                 123.81
No. Observations:                 244   AIC:                            -243.6
Df Residuals:                     242   BIC:                            -236.6
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          0.6181      0.054     11.444      0.000       0.512       0.724
net_bill      -0.1534      0.020     -7.820      0.000      -0.192      -0.115
==============================================================================
Omnibus:                      440.899   Durbin-Watson:                   1.965
Prob(Omnibus):                  0.000   Jarque-Bera (JB):           171509.052
Skew:                           9.850   Prob(JB):                         0.00
Kurtosis:                     131.381   Cond. No.                         17.9
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

R-quadrado: 0.202. Isso significa que cerca de 20.2% da variabilidade em tip_pct é explicada pelo logaritmo de net_bill. Esse modelo apresenta uma melhoria em relação ao modelo anterior sem logaritmo.

Intercepto (constante): 0.6181. Indica que, se o logaritmo de net_bill fosse 0, o percentual esperado da gorjeta seria aproximadamente 61.81%.

Coeficiente de net_bill (logarítmico): -0.1534. Este coeficiente negativo sugere que, à medida que o valor de net_bill aumenta, o percentual da gorjeta diminui, mesmo após a transformação logarítmica.

###  3. ajuste o modelo de ```tip_pct``` em função de um polinômio de segundo grau em ```net_bill```.

In [18]:
poly = PolynomialFeatures(degree=2, include_bias=False)
X_poly = poly.fit_transform(tips_df[['net_bill']])

X_poly = sm.add_constant(X_poly)

model_poly = sm.OLS(y, X_poly).fit()

model_poly.summary()


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                tip_pct   R-squared:                       0.158
Model:                            OLS   Adj. R-squared:                  0.151
Method:                 Least Squares   F-statistic:                     22.55
Date:                Mon, 06 May 2024   Prob (F-statistic):           1.06e-09
Time:                        21:18:02   Log-Likelihood:                 117.25
No. Observations:                 244   AIC:                            -228.5
Df Residuals:                     241   BIC:                            -218.0
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          0.4851      0.047     10.332      0.000       0.393       0.578
x1            -0.0268      0.005     -5.477      0.000      -0.036      -0.017
x2             0.0005      0.000      4.356      0.000       0.000       0.001
==============================================================================
Omnibus:                      447.620   Durbin-Watson:                   2.026
Prob(Omnibus):                  0.000   Jarque-Bera (JB):           186035.672
Skew:                          10.140   Prob(JB):                         0.00
Kurtosis:                     136.744   Cond. No.                     2.42e+03
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 2.42e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

R-quadrado representa uma melhoria em relação ao modelo linear simples, indicando que aproximadamente 15.8% da variabilidade em 'tip_pct' é explicada pelo modelo polinomial de 'net_bill'.

###  4. ajuste um modelo no log do ```tip_pct``` em função de ```net_bill```. Calcule o $R^2$ também em ```tip_pct```

In [19]:
y_log = np.log(tips_df['tip_pct'])  # Transformação logarítmica
X = tips_df[['net_bill']]           # Variável independente
X = sm.add_constant(X)              # Adicionar uma constante ao modelo

model_log = sm.OLS(y_log, X).fit()

y_pred_log = model_log.predict(X)
y_pred = np.exp(y_pred_log)  # Aplicar a transformação inversa do logaritmo

# Calcular R-quadrado manualmente para a transformação inversa
ss_res = np.sum((tips_df['tip_pct'] - y_pred) ** 2)
ss_tot = np.sum((tips_df['tip_pct'] - np.mean(tips_df['tip_pct'])) ** 2)
r_squared = 1 - (ss_res / ss_tot)

model_log.summary(), r_squared


(<class 'statsmodels.iolib.summary.Summary'>
 """
                             OLS Regression Results                            
 Dep. Variable:                tip_pct   R-squared:                       0.235
 Model:                            OLS   Adj. R-squared:                  0.232
 Method:                 Least Squares   F-statistic:                     74.48
 Date:                Mon, 06 May 2024   Prob (F-statistic):           8.27e-16
 Time:                        21:21:14   Log-Likelihood:                -106.00
 No. Observations:                 244   AIC:                             216.0
 Df Residuals:                     242   BIC:                             223.0
 Df Model:                           1                                         
 Covariance Type:            nonrobust                                         
                  coef    std err          t      P>|t|      [0.025      0.975]
 ---------------------------------------------------------------------

###  5. ajuste um modelo no log do ```tip_pct``` em função do log de ```net_bill```. Calcule o $R^2$ também em ```tip_pct```.

In [20]:
X_log = np.log(tips_df['net_bill'])  
X_log = sm.add_constant(X_log)    

model_log_log = sm.OLS(y_log, X_log).fit()

y_pred_log_log = model_log_log.predict(X_log)
y_pred_log = np.exp(y_pred_log_log)  # Aplicar a transformação inversa do logaritmo

ss_res_log = np.sum((tips_df['tip_pct'] - y_pred_log) ** 2)
ss_tot_log = np.sum((tips_df['tip_pct'] - np.mean(tips_df['tip_pct'])) ** 2)
r_squared_log = 1 - (ss_res_log / ss_tot_log)

model_log_log.summary(), r_squared_log


(<class 'statsmodels.iolib.summary.Summary'>
 """
                             OLS Regression Results                            
 Dep. Variable:                tip_pct   R-squared:                       0.293
 Model:                            OLS   Adj. R-squared:                  0.291
 Method:                 Least Squares   F-statistic:                     100.5
 Date:                Mon, 06 May 2024   Prob (F-statistic):           5.20e-20
 Time:                        21:23:26   Log-Likelihood:                -96.357
 No. Observations:                 244   AIC:                             196.7
 Df Residuals:                     242   BIC:                             203.7
 Df Model:                           1                                         
 Covariance Type:            nonrobust                                         
                  coef    std err          t      P>|t|      [0.025      0.975]
 ---------------------------------------------------------------------

###  6. qual modelo resultou no melhor ajuste?

Entre os modelos ajustados, o modelo que utilizou o logaritmo de 'tip_pct' em função do logaritmo de 'net_bill' resultou no melhor ajuste.

O modelo Log-Log não só apresentou o maior R-squared (0.293) para os dados na escala logarítmica, como também, ao converter de volta para a escala original de 'tip_pct', manteve um R-squared relativamente alto (0.228).